filebeat/input/file/states.go

```diff
@@ -0,0 +1,157 @@
+package file
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// States handles list of FileState. One must use NewStates to instantiate a
+// file states regisry. Using the zero-value is not safe.
+type States struct {
+	sync.RWMutex
+
+	// states store
+	states []State
+
+	// idx maps state IDs to state indexes for fast lookup and modifications.
+	idx map[string]int
+}
+
+// NewStates generates a new states registry.
+func NewStates() *States {
+	return &States{
+		states: nil,
+		idx:    map[string]int{},
+	}
+}
+
+// Update updates a state. If previous state didn't exist, new one is created
+func (s *States) Update(newState State) {
+	s.UpdateWithTs(newState, time.Now())
+}
+
+// UpdateWithTs updates a state, assigning the given timestamp.
+// If previous state didn't exist, new one is created
+func (s *States) UpdateWithTs(newState State, ts time.Time) {
+	s.Lock()
+	defer s.Unlock()
+
+	id := newState.ID()
+	index := s.findPrevious(id)
+	newState.Timestamp = ts
+
+	if index >= 0 {
+		s.states[index] = newState
+	} else {
+		// No existing state found, add new one
+		s.idx[id] = len(s.states)
+		s.states = append(s.states, newState)
+		logp.Debug("input", "New state added for %s", newState.Source)
+	}
+}
+
+// FindPrevious lookups a registered state, that matching the new state.
+// Returns a zero-state if no match is found.
+func (s *States) FindPrevious(newState State) State {
+	s.RLock()
+	defer s.RUnlock()
+	i := s.findPrevious(newState.ID())
+	if i < 0 {
+		return State{}
+	}
+	return s.states[i]
+}
+
+// findPrevious returns the previous state for the file.
+// In case no previous state exists, index -1 is returned
+func (s *States) findPrevious(id string) int {
+	if i, exists := s.idx[id]; exists {
+		return i
+	}
+	return -1
+}
+
+// Cleanup cleans up the state array. All states which are older then `older` are removed
+// The number of states that were cleaned up and number of states that can be
+// cleaned up in the future is returned.
+func (s *States) Cleanup() (int, int) {
+	s.Lock()
+	defer s.Unlock()
+
+	currentTime := time.Now()
+	statesBefore := len(s.states)
+	numCanExpire := 0
+
+	L := len(s.states)
+	for i := 0; i < L; {
+		state := &s.states[i]
+		canExpire := state.TTL > 0
+		expired := (canExpire && currentTime.Sub(state.Timestamp) > state.TTL)
+
+		if state.TTL == 0 || expired {
+			if !state.Finished {
+				logp.Err("State for %s should have been dropped, but couldn't as state is not finished.", state.Source)
+				i++
+				continue
+			}
+
+			delete(s.idx, state.ID())
+			logp.Debug("state", "State removed for %v because of older: %v", state.Source, state.TTL)
+
+			L--
```

Comment:
Looking at the code with `i++` and `L--` I'm thinking if the option with combining index and registry would actually bring less complexity?

Reply:
Maybe it's not that clear as some extra work is done within the loop, but the loop basically sorts the list into alive and dead states. The L is the current length. Removal/sorting is in-place. That is, we're forced to check the current index again upon removal.
The `i++` and `L--` act as delimiters of processed and alive/dead states => both need to be updated, depending if state is alive or not. Very basic in-place removal loop without extra copies.

would this be more clear for you:

```
func (s *States) Cleanup() (int, int) {
	s.Lock()
	defer s.Unlock()

	L := len(s.states)
	currentTime := time.Now()
	statesBefore := L
	numCanExpire := 0 // number of alive states with TTL > 0

	for i := 0; i < L; {
		state := &s.states[i]
		canExpire := state.TTL > 0
		expired := (canExpire && currentTime.Sub(state.Timestamp) > state.TTL)

		if (state.TTL == 0 || expired) && state.Finished {
			// remove by replacing entry at i with last entry and shrink 'alive' states length
			L--
			if L != i {
				// swap and update index
				s.states[i], s.states[L] = s.states[L], s.states[i]
				s.idx[s.states[i].ID()] = i
			}
		} else {
			i++
			if canExpire {
				numCanExpire++
			}
		}
	}

	alive := s.states[:L]
	dead := s.states[L:]
	for i := range dead {
		delete(s.idx, dead[i].ID())
	}

	s.states = alive
	return statesBefore - len(s.states), numCanExpire
}
```

Reply:
When I've looked at the code initially I had to use pseudo code here to make sure the in place was working correctly. 

I prefer the first version.